### PR TITLE
Changed implementation of `genbank2uid`

### PR DESCRIPTION
**Pull request by zachary-foster**
## Description

`genbank2uid` now returns the correct ID when there are multiple possibilities and invalid IDs no longer make whole batches fail. I used `esummary` instead of `elink` to look up taxon ids and parsed the error messages to see which queries failed. In the previous implementation, `esummary` was called anyway, so this should be a bit faster.  

## Related Issue

This might finally solve issue #506

Also https://github.com/grunwaldlab/metacoder/issues/177

## Example

```
> genbank2uid(c("NC_002737", "nope", "NC_013791.2", "alsonope", "15638860"))
[[1]]
[1] "160490"
attr(,"class")
[1] "uid"
attr(,"match")
[1] "found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/160490"
attr(,"name")
[1] "Streptococcus pyogenes M1 GAS, complete genome"

[[2]]
[1] NA
attr(,"class")
[1] "uid"
attr(,"match")
[1] "not found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/NA"
attr(,"name")
[1] "unknown"

[[3]]
[1] "398511"
attr(,"class")
[1] "uid"
attr(,"match")
[1] "found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/398511"
attr(,"name")
[1] "Bacillus pseudofirmus OF4, complete genome"

[[4]]
[1] NA
attr(,"class")
[1] "uid"
attr(,"match")
[1] "not found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/NA"
attr(,"name")
[1] "unknown"

[[5]]
[1] "9606"
attr(,"class")
[1] "uid"
attr(,"match")
[1] "found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/9606"
attr(,"name")
[1] "Homo sapiens BAC clone RP11-460N20 from 7, complete sequence"

Warning message:
The following 2 of 5 queries could not be found:
   [2] nope, [4] alsonope
```

### Tests

I tried to replicate the output format I saw before making the changes, but it seems the tests expect a character vector instead of a list? The tests fail, but as far as I can tell they would have failed before this change.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -41,6 +41,8 @@ getting simplified output (#623) thanks @glaroc !
 
 ### BUG FIXES
 
+* `genbank2uid` now returns the correct ID when there are multiple possibilities
```
Comment (sckott) [inline NEWS.md]: don't add a news item in a version that's already on CRAN - can start a new section with a dev version number or so

Comment (zachary-foster) [reply NEWS.md]: Oops, I assumed that the top section was the dev version. Should have checked that..

**Comment by sckott** [comment]
this `genbank2uid(232323232323, config=verbose())` seems to be failing

Maybe cause the error message is spread across many lines? 

```
curl 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=nucleotide&id=232323232323'
```

```
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE eSummaryResult PUBLIC "-//NLM//DTD esummary v1 20041029//EN" "https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20041029/esummary-v1.dtd">
<eSummaryResult>
<ERROR>Iderror uid="232323232323" term="232323232323" function="GetGiClass"</ERROR>
<ERROR>exception uid="232323232323" MakeDocsum can't get data from PUBSEQ_OS (original msg: Can't get Blob Flags - gi is not found).</ERROR>
<ERROR>Failed uid="232323232323"</ERROR>
```

**Comment by zachary-foster** [comment]
Hmm, looks like there are multiple ways the errors are reported. Check out:

https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=nucleotide&id=NC_002737,nope,NC_013791,alsonope,232323232323

I can still parse it, but it will get cumbersome if there are too many other types of errors.

**Comment by sckott** [comment]
it's truly a hot 🔥 mess 

I guess do as good as you can without getting too complicated

File: R/genbank2uid.R [hunk 2]
```diff
@@ -39,61 +34,78 @@ genbank2uid <- function(id, batch_size = 100, ...) {
...
+    # Extract taxon ID and sequences name
+    taxon_ids <- xml_text(xml_find_all(parsed_xml,
+                                       "//eSummaryResult//DocSum//Item[@Name='TaxId']"))
+    titles <- xml_text(xml_find_all(parsed_xml,
+                                    "//eSummaryResult//DocSum//Item[@Name='Title']"))
+    
+    # Add NAs for failed queries
+    raw_errors <- xml_text(xml_find_all(parsed_xml, "//eSummaryResult//ERROR"))
+    if (length(raw_errors) > 0) {
+      error_regexes <- c('^Invalid uid (.+) at position=[0-9]+$',
```
Comment (zachary-foster) [inline R/genbank2uid.R]: We can add regexes to this if more error types pop up.

Hmm, I dont know how to use this review tool right. The variable `error_regexes` was what I was talking about.

**Comment by zachary-foster** [comment]
Now we get: 

```
> genbank2uid(c("15638860", "232323232323"))
[[1]]
[1] "9606"
attr(,"class")
[1] "uid"
attr(,"match")
[1] "found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/9606"
attr(,"name")
[1] "Homo sapiens BAC clone RP11-460N20 from 7, complete sequence"

[[2]]
[1] NA
attr(,"class")
[1] "uid"
attr(,"match")
[1] "not found"
attr(,"multiple_matches")
[1] FALSE
attr(,"pattern_match")
[1] FALSE
attr(,"uri")
[1] "http://www.ncbi.nlm.nih.gov/taxonomy/NA"
attr(,"name")
[1] "unknown"

Warning message:
The following 1 of 2 queries could not be found:
   [2] 232323232323
```

**Comment by sckott** [comment]
LGTM

**Comment by zachary-foster** [comment]
Cool! Thanks

**Pull request merged**